File: onibus/src/main/java/com/proj_db/onibus/service/MotorService.java
package com.proj_db.onibus.service;

import java.util.List;
import java.util.Optional;

import com.proj_db.onibus.dto.MotorCreateDTO;
import com.proj_db.onibus.dto.MotorUpdateDTO;
import com.proj_db.onibus.model.Motor;

public interface MotorService {
    
    // --- CRUD Básico ---
    Motor save(MotorCreateDTO motor);
    Motor update(Long id, MotorUpdateDTO motorDetails);
    void deleteById(Long id);
    Optional<Motor> findById(Long id);
    List<Motor> findAll();

    // --- Buscas Específicas ---
    Optional<Motor> findByNumeroSerie(String numeroSerie);
    List<Motor> search(MotorSearchDTO criteria);

    // --- Lógica de Negócio (Ciclo de Vida) ---
    Motor enviarParaManutencao(Long motorId);
    Motor retornarDaManutencao(Long motorId);
    Motor enviarParaRevisao(Long motorId);
    Motor retornarDaRevisao(Long motorId);

    // --- Lógica de OS Preventiva ---
    void verificarEGerarOsPreventivas();

    // --- Métodos de Relatório ---
    List<Object[]> countByTipo();
    List<Object[]> countByStatus();
    
    // DTO (Data Transfer Object) para a busca
    record MotorSearchDTO(
        String marca,
        String modelo,
        String numeroSerie,
        String tipoOleo,
        Motor.TipoMotor tipo,
        Motor.StatusMotor status,
        Integer potenciaMin,
        Integer potenciaMax
    ) {}
}

File: frontend-onibus/src/services/motorService.js
import api from './api';

/**
 * Objeto que encapsula todas as chamadas de API para a entidade 'Motor'.
 */
export const motorService = {
  
  // --- Métodos CRUD ---
  create: (motorCreateDTO) => api.post('/motores', motorCreateDTO),
  update: (id, motorUpdateDTO) => api.put(`/motores/${id}`, motorUpdateDTO),
  delete: (id) => api.delete(`/motores/${id}`),
  
  // --- Consultas ---
  getAll: () => api.get('/motores'),
  getById: (id) => api.get(`/motores/${id}`),
  search: (terms) => api.get('/motores/search', { params: terms }),

  // --- Ações de Negócio (Ciclo de Vida) ---
  enviarParaManutencao: (motorId) => api.patch(`/motores/${motorId}/enviar-manutencao`),
  retornarDeManutencao: (motorId) => api.patch(`/motores/${motorId}/retornar-manutencao`),
  enviarParaRevisao: (motorId) => api.patch(`/motores/${motorId}/enviar-revisao`),
  retornarDaRevisao: (motorId) => api.patch(`/motores/${motorId}/retornar-revisao`),
};
